### PR TITLE
Upload Introspection Query

### PR DESCRIPTION
**Pull request by jamestouri**
https://github.com/medplum/medplum/issues/1937

### CODE REVIEW

File: scripts/deploy-introspection-schema.sh [hunk 1]
```diff
@@ -0,0 +1,24 @@
+# Taking a compressed introspection schema file and uploading it to S3
+# This is specifically for s3://graphiq.medplum.com/schema/
+
+version=$1
+client_id=$2
+client_secret=$3
+cors_file=$4
```
Comment (codyebberson) [inline scripts/deploy-introspection-schema.sh:7]: What goes into `cors_file`?  If it's constant, can we include it here in the bash script?

File: scripts/deploy-introspection-schema.sh [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
...tion            fields(includeDeprecated: true) {        name        description        args {          ...InputValue        }        type {          ...TypeRef        }        isDeprecated        deprecationReason      }      inputFields {        ...InputValue      }      interfaces {        ...TypeRef      }      enumValues(includeDeprecated: true) {        name        description        isDeprecated        deprecationReason      }      possibleTypes {        ...TypeRef      }    }    fragment InputValue on __InputValue {      name      description      type { ...TypeRef }      defaultValue                }    fragment TypeRef on __Type {      kind      name      ofType {        kind        name        ofType {          kind          name          ofType {            kind            name            ofType {              kind              name              ofType {                kind                name                ofType {                  kind                  name                  ofType {                    kind                    name                  }                }              }            }          }        }      }    }  ","operationName":"IntrospectionQuery"}' \
```
Comment (codyebberson) [inline scripts/deploy-introspection-schema.sh:12]: Super pedantic nit, but can we collapse all of the whitespace?  You can use a regex replace in VS Code:

Search: `\s+`
Replace: ` ` (a single space)

File: scripts/deploy-introspection-schema.sh [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+gzip schema-$version.json
+
+aws s3api put-bucket-cors \
+  --bucket graphiq.medplum.com \
+  --cors-configuration $cors_file
+
+aws s3 cp ../schema-$version.json.gz "s3://graphiq.medplum.com/schema/" \
+  --content-type "application/json" \
+  --content-encoding "gzip" \
+  --cache-control "public, max-age=31536000"
```
Comment (codyebberson) [inline scripts/deploy-introspection-schema.sh:24]: Add new line at end of file

File: scripts/deploy-introspection-schema.sh [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  > schema-$version.json
+
+gzip schema-$version.json
+
+aws s3api put-bucket-cors \
```
Comment (codyebberson) [inline scripts/deploy-introspection-schema.sh:17]: Is this a one-time thing? Or do we need to call this everytime we update the file?

If it's a one time thing, then maybe we should pull this out of the script entirely.

Comment (jamestouri) [reply scripts/deploy-introspection-schema.sh:17]: It's a one-time thing

File: scripts/deploy-introspection-schema.sh [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+curl 'http://localhost:8103/fhir/R4/$graphql' \
+  -u $client_id:$client_secret \
+  -H 'Content-Type: application/json' \
+  --data-raw '{"query":" query IntrospectionQuery { __schema { queryType { name } mutationType { name } subscriptionType { name } types { ...FullType } directives { name description locations args { ...InputValue } } } } fragment FullType on __Type { kind name description fields(includeDeprecated: true) { name description args { ...InputValue } type { ...TypeRef } isDeprecated deprecationReason } inputFields { ...InputValue } interfaces { ...TypeRef } enumValues(includeDeprecated: true) { name description isDeprecated deprecationReason } possibleTypes { ...TypeRef } } fragment InputValue on __InputValue { name description type { ...TypeRef } defaultValue } fragment TypeRef on __Type { kind name ofType { kind name ofType { kind name ofType { kind name ofType { kind name ofType { kind name ofType { kind name ofType { kind name } } } } } } } } ","operationName":"IntrospectionQuery"}' \
+  > schema-$version.json
+
+gzip schema-$version.json
+
+aws s3 cp ../schema-$version.json.gz "s3://graphiq.medplum.com/schema/" \
```
Comment (codyebberson) [inline scripts/deploy-introspection-schema.sh:16]: The resulting file should not have the `.gz` extension.  I think we should be able to do that by specifying the file name in the command here:

```bash
aws s3 cp "../schema-$version.json.gz" "s3://graphiq.medplum.com/schema/schema-$version.json"
```

Also, I assume it should only be one period ".", not two periods "..", for current directory?

File: scripts/deploy-introspection-schema.sh [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+  --content-type "application/json" \
+  --content-encoding "gzip" \
+  --cache-control "public, max-age=31536000"
+
+secret_value="/schema/schema-$version.json.gz"
```
Comment (codyebberson) [inline scripts/deploy-introspection-schema.sh:21]: Remove the `.gz` from the filename.

File: scripts/deploy-introspection-schema.sh [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+curl 'http://localhost:8103/fhir/R4/$graphql' \
+  -u $client_id:$client_secret \
+  -H 'Content-Type: application/json' \
+  --data-raw '{"query":" query IntrospectionQuery { __schema { queryType { name } mutationType { name } subscriptionType { name } types { ...FullType } directives { name description locations args { ...InputValue } } } } fragment FullType on __Type { kind name description fields(includeDeprecated: true) { name description args { ...InputValue } type { ...TypeRef } isDeprecated deprecationReason } inputFields { ...InputValue } interfaces { ...TypeRef } enumValues(includeDeprecated: true) { name description isDeprecated deprecationReason } possibleTypes { ...TypeRef } } fragment InputValue on __InputValue { name description type { ...TypeRef } defaultValue } fragment TypeRef on __Type { kind name ofType { kind name ofType { kind name ofType { kind name ofType { kind name ofType { kind name ofType { kind name ofType { kind name } } } } } } } } ","operationName":"IntrospectionQuery"}' \
+  > schema-$version.json
+
+gzip schema-$version.json
+
+aws s3 cp ./schema-$version.json "s3://graphiq.medplum.com/schema/schema-$version.json" \
```
Comment (codyebberson) [inline scripts/deploy-introspection-schema.sh:16]: Does this work?  I assume that the output of line 14 will be a new file called `schema-$version.json.gz`, so this command would need to use that filename?

**Pull request merged**